src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -290,8 +328,48 @@ public override int Read(byte[] buffer, int offset, int count)
             }
         }
 
+        private static async Task<int> ReadConditionallyAsync(Stream s, byte[] buffer, int offset, int count, CancellationToken cancellationToken, bool useAsync)
+        {
+            if (useAsync)
+                return await s.ReadAsync(buffer, offset, count, cancellationToken);
+            else
+                return s.Read(buffer, offset, count);
+        }
```

Comment:
I can see why this helper is tempting, and functionally it's fine, but it's going to add more overhead than you'd expect.  If you really want it as a helper, it'd be better to write it as a non-async method, e.g.
```C#
private static Task<int> ReadConditionallyAsync(Stream s, byte[] buffer, int offset, int count, CancellationToken cancellationToken, bool useAsync)
{
    return useAsync ?
        s.ReadAsync(buffer, offset, count, cancellationToken) :
        Task.FromResult(s.Read(buffer, offset, count);
}
```
as that'll have fewer allocations (just the extra one for the Task.FromResult in the synchronous case) than the previous version (which will likely have at least a Task allocation, plus several additional allocations if the async operation doesn't complete synchronously).

But... you can avoid all of those if you just don't have the helper, and the two places you need it just writing it out, e.g.
```
amountRead += useAsync ?
    await _stream.ReadAsync(tempInputBuffer, _inputBufferIndex, numWholeBlocksInBytes - _inputBufferIndex, cancellationToken) :
    _stream.Read(tempInputBuffer, _inputBufferIndex, numWholeBlocksInBytes - _inputBufferIndex);
```